Autocomplete for starter code repo

cyhsutw opened the pull request:
Hello guys,

This should fix #480. 

Here's a short demo:

![autocomplete](https://cloud.githubusercontent.com/assets/1046917/13902654/f6529286-ee8b-11e5-80d8-91874d8db628.gif)

- By default, it shows the repositories that the user can access using the [list your repos API](https://developer.github.com/v3/repos/#list-your-repositories).
- Or, if the user put type some string like `education/t`, it will search the the repositories under namespace education (in this case, it will show `education/teachers_pet` and `education/teachers`).
- It utilizes the [repository search API](https://developer.github.com/v3/search/#search-repositories) directly for other queries like `rails`.

### Issues with Testing

Previously, I have the `AutocompleteController` generate JSON responses and use CoffeeScript to create suggestion elements.
After digging into the code base, I found using the `render_partial` might be more extensible for future additions like searching `users`.

After this change, it becomes hard to write tests since we will deal with the HTML contents which is relatively harder to validate.

One possible solution is making the controller support both HTML and JSON and the tests will check the JSON responses instead of HTML responses. 

Is there any good way to make the `AutocompleteController` more testable?

------

I would love to participate GSoC 2016 and this is my first PR for the Classroom. I would be very grateful if you can provide any advice/feedback :smile: 

Thank you!

tarebyte commented on config/routes.rb:
```diff
@@ -12,6 +12,9 @@
   match '/auth/:provider/callback', to: 'sessions#create',  via: [:get, :post]
   match '/auth/failure',            to: 'sessions#failure', via: [:get, :post]
 
+  # autocomplete
```
This comment isn't really needed, would you mind taking it out?

tarebyte commented on app/views/group_assignments/_group_assignment_form_options.html.erb:
```diff
@@ -19,7 +19,10 @@
-  <dd><input class="textfield" type="text" name="repo_name" value="<%= group_assignment.full_name %>" placeholder="<%= group_assignment.full_name || 'e.g. education/classroom'%>"</dd>
+  <dd><input id="autocomplete-textfield" class="textfield" type="text" name="repo_name" value="<%= group_assignment.full_name %>" placeholder="<%= group_assignment.full_name || 'e.g. education/classroom'%>"></dd>
+
+  <!-- autocomplete suggestions -->
```
I'm not really a big fan of comments in the markup could you also remove this comment?

tarebyte commented on app/views/assignments/_assignment_form_options.html.erb:
```diff
@@ -19,5 +19,8 @@
-  <dd><input class="textfield" type="text" name="repo_name" value="<%= assignment.full_name %>" placeholder="<%= assignment.full_name || 'e.g. education/classroom'%>"</dd>
+  <dd><input id="autocomplete-textfield" class="textfield" type="text" name="repo_name" value="<%= assignment.full_name %>" placeholder="<%= assignment.full_name || 'e.g. education/classroom'%>" autocomplete="off"></dd>
+
+  <!-- autocomplete suggestions -->
```
Ditto https://github.com/education/classroom/pull/508/files#r56770268

tarebyte commented on app/assets/stylesheets/application.scss:
```diff
@@ -14,3 +14,6 @@
 
 @import "remodal/dist/remodal";
+
+// autocomplete
```
The `@import` is pretty self explanatory here, would mind removing this comment?

tarebyte commented:
@cyhsutw would you mind using this spinner instead?

![octocat-spinner-32](https://cloud.githubusercontent.com/assets/564113/13906695/93b2f300-eeb3-11e5-8958-bd73098eb329.gif)

tarebyte commented on app/assets/javascripts/autocomplete.coffee:
```diff
@@ -0,0 +1,69 @@
+  # TODO: test with IE 8
+  # onpropertychange -> oninput for IE 8
+  # https://msdn.microsoft.com/en-us/library/ms536956(v=vs.85).aspx
+  $('#autocomplete-textfield').on('focus input propertychange', ->
```
GitHub doesn't support IE 8 https://help.github.com/articles/supported-browsers/, so we don't have to worry about testing for that.

tarebyte replied:
I'm also not a huge fan of `id` based javascript bindings.

Generally what I do is add another class to the DOM object and add a `js-` prefix to it.

Example.

```html
<dd><input class="textfield js-autocomplete-textfield" ...>...</dd>
```

So then your coffee would look like
```coffee
$('.js-autocomplete-textfield').on('focus input propertychange', ->
```

tarebyte commented on app/controllers/autocomplete_controller.rb:
```diff
@@ -0,0 +1,28 @@
+    end
+    current_user.github_search_client.search_repos(keyword, per_page: 10, page: 1)[:items]
+  end
+end
```
So I think this controller can be cleaned up a little bit, we can remove some of the logic away from the `#search_repos` action.

```ruby
class AutocompleteController < ApplicationController
  def search_repos
    repos = search_github_repos(params[:query])

    respond_to do |format|
      format.html { render partial: 'autocomplete/repository_suggestions', locals: { repos: repos } }
    end
  end

  private

  def github_search_client
    @github_search_client ||= current_user.github_search_client
  end

  def search_github_repos(query)
    return github_search_client.repos(nil, sort: 'updated', per_page: 10, page: 1) unless query.present?

    keyword = query.gsub(%r{^[a-zA-Z0-9_-]+\/}, '') + ' in:name fork:true'

    # add namespace criteria if needed
    if query.include? '/'
      namespace = query.split('/').first
      keyword += " user:#{namespace}"
    end

    github_search_client.search_repos(keyword, per_page: 10, page: 1)[:items]
  end
end
```

tarebyte commented on config/routes.rb:
```diff
@@ -12,6 +12,9 @@
   match '/auth/failure',            to: 'sessions#failure', via: [:get, :post]
 
+  # autocomplete
+  get '/autocomplete/repo_name', to: 'autocomplete#search_repos'
```
What do you think about

```ruby
get '/autocomplete/github_repos', to: 'autocomplete#search_github_repos'
```

tarebyte commented on app/assets/javascripts/autocomplete.coffee:
```diff
@@ -0,0 +1,69 @@
+    update_textfield(selected_items.first()) if selected_items.length == 1
+
+    # hide suggestion list
+    $('#autocomplete-suggestions-container').hide()
```
All of these comments aren't needed, the code gives a pretty clear picture of what is happening.

tarebyte commented on app/assets/stylesheets/autocomplete.scss:
```diff
@@ -0,0 +1,71 @@
+  overflow-x: hidden;
+  overflow-y: scroll;
+
+  // list
```
This comment isn't really needed, it doesn't add anymore context to what is going on.

tarebyte commented on app/assets/stylesheets/autocomplete.scss:
```diff
@@ -0,0 +1,71 @@
+  ul {
+    list-style: none;
+
+    // item
```
ditto https://github.com/education/classroom/pull/508/files#r56771369

tarebyte commented on app/assets/stylesheets/autocomplete.scss:
```diff
@@ -0,0 +1,71 @@
+    }
+  }
+
+  // octicon
```
ditto https://github.com/education/classroom/pull/508/files#r56771369

tarebyte commented on app/assets/stylesheets/autocomplete.scss:
```diff
@@ -0,0 +1,71 @@
+  }
+
+  // spinner
+  .autocomplete-loading-indicator {
```
ditto https://github.com/education/classroom/pull/508/files#r56771369

tarebyte commented on app/assets/stylesheets/autocomplete.scss:
```diff
@@ -0,0 +1,71 @@
+
+
+// blank status
+.autocomplete-suggestions-empty {
```
https://github.com/education/classroom/pull/508/files#r56771369

tarebyte commented on app/assets/javascripts/autocomplete.coffee:
```diff
@@ -0,0 +1,69 @@
+        # the `onBlur` event will take place before `onClick`.
+        # So when a user hovers the item, we can mark that item first
+        # and retrieve the hovered item when `onBlur` is fired.
+        $('.autocomplete-suggestion-item').hover(->
```
Can we use a `js-` prefix class here so we don't mix javascript with css styling?

tarebyte commented on app/assets/javascripts/autocomplete.coffee:
```diff
@@ -0,0 +1,69 @@
+        $('.autocomplete-suggestion-item').hover(->
+          $(this).addClass('suggestion-focused')
+        , ->
+          $(this).removeClass('suggestion-focused')
```
You defined `$(this)` here https://github.com/education/classroom/pull/508/files#diff-a34de611c45045bee0ea980cd2c0471cR19, but it looks like it never gets used.

tarebyte commented:
@cyhsutw this is looking really good so far, I left some comments.

Regarding the testing we don't really do any markup testing right now.

What I would do is make controller tests that verify that you get search results from the controller action.

cyhsutw commented:
Hi @tarebyte thanks for reviewing my code.

Your comments are really helpful for me to improve my coding styles.

I've updated my code according to your suggestions, please review them when you are available.

Thank you! :smile:

cyhsutw commented on app/views/assignments/_assignment_form_options.html.erb:
```diff
@@ -19,5 +19,6 @@
 <dl class="form">
   <dt><label>Add your starter code from GitHub <small>(optional)</small> </label></dt>
-  <dd><input class="textfield" type="text" name="repo_name" value="<%= assignment.full_name %>" placeholder="<%= assignment.full_name || 'e.g. education/classroom'%>"</dd>
+  <dd><input class="textfield js-autocomplete-textfield" type="text" name="repo_name" value="<%= assignment.full_name %>" placeholder="<%= assignment.full_name || 'e.g. education/classroom'%>" autocomplete="off" data-autocomplete-search-endpoint="github_repos"></dd>
```
Previously, we used the `name` attribute to determine the endpoint to send request to. 

As we changed the `routes.rb`, I then add a custom [`data-*`](https://www.w3.org/TR/html5/dom.html#embedding-custom-non-visible-data-with-the-data-*-attributes) attribute to specify the endpoint.

johndbritton commented on app/assets/javascripts/autocomplete.coffee:
```diff
@@ -0,0 +1,54 @@
+
+update_textfield = (list_element) ->
+  $(list_element).removeClass('suggestion-focused')
+  $('.js-autocomplete-textfield').val($(list_element).find('.res-name')[0].innerHTML)
```
What do you think about modifying the form to submit `repository_id` when using autocompletion instead of sending `name/owner`?

I know it's an unlikely case, but it's possible that a user could rename the repository in between the time of choosing the item from the list and submitting the page. By using `repository_id` we avoid that

cyhsutw replied:
:+1: 

This is a possible case, I would change the name if I am not satisfied with it :stuck_out_tongue:

What do you think about adding a hidden `input` with `repository_id` as `value` and we can use the value if it's present?

johndbritton replied:
And it also just feels cleaner to reference by `repository_id`

johndbritton replied:
> What do you think about adding a hidden input with repository_id as value and we can use the value if it's present?

Yeah, I like this idea. By default send `owner/name` but if it's selected via autocomplete also send a hidden `repository_id` and use that if it's supplied. You're going to need to make sure you have tests for this as well.

I'd check with @tarebyte to make sure he likes this plan before implementing.

tarebyte replied:
That sounds fine to me, we have validations that the repo belongs to the user on the controller so I'm not worried about anyone spoofing us.

johndbritton commented on app/models/user.rb:
```diff
@@ -39,6 +39,11 @@ def github_client_scopes
   end
 
+  # for searching, disable pagination so we can control the number of suggestions
+  def github_search_client
```
I don't love that we're adding another method to the `User` here. Why is this the right place for this code?

tarebyte replied:
I think the reasoning here is because we need the users `access_token` to view their private repositories

johndbritton replied:
Yeah, I see that we need the access token from the user, but why does this client belong on the user? Couldn't you also do something in the controller like this

```ruby
def search_client
  Octokit::Client.new(access_token: current_user.token, auto_paginate: false)
end
```

I'm not sure why, but I don't really like adding this method here to use it in just one case. Alternatively, what about adding an options hash argument to `User#github_client` that would allow you to disable pagination with that method?

cyhsutw replied:
Actually we can access the `token` attribute outside the class, so we can create the client in the controller.

My concerns were about concurrency, but I forgot that HTTP is stateless :sweat_smile:

> Alternatively, what about adding an options hash argument to User#github_client that would allow you to disable pagination with that method?

I thinks this is more reasonable :+1:

tarebyte replied:
@johndbritton that sound fine to me

@cyhsutw if you could memoize the `seach_client` that'd be ✨ 

```ruby
def search_client
  @search_client ||= Octokit::Client.new(access_token: current_user.token, auto_paginate: false)
end
```

cyhsutw replied:
@johndbritton @tarebyte 

Hi @johndbritton 

What about declaring `github_client` like this:

```ruby
def github_client(opts = {})
  @github_client ||= Octokit::Client.new(opts.merge(access_token: token, auto_paginate: true))
end
```

Then we can have our `search_client` like this:

```ruby
def search_client
  @search_client ||= current_user.github_client(auto_paginate: false)
end
```

cyhsutw replied:
Oops! Please ignore my comment :arrow_up: 

Looks like it doesn't work since we have conditional assignment on `@github_client`, which is assigned before the `search_client` is called.

cyhsutw commented:
Hi @johndbritton @tarebyte 

I've updated the code based on your suggestions, please review them when you are available.

Thank you! :smile:

cyhsutw commented:
> What I would do is make controller tests that verify that you get search results from the controller action.

@tarebyte does this mean that we only test if the endpoint works?

I tried to write some tests for the `AutocompleteController`, here's what I got so far:

```ruby
RSpec.describe AutocompleteController, type: :controller do
  let(:organization) { GitHubFactory.create_owner_classroom_org }
  let(:user)         { organization.users.first                 }

  before do
    sign_in(user)
  end

  describe 'GET #github_repos', :vcr do
    it 'returns success and returns repos of the user' do
      get :github_repos
      expect(response).to have_http_status(:success)
      expect(response).to render_template(partial: 'autocomplete/_repository_suggestions')
    end

    it 'returns success and returns search results of a query' do
      get :github_repos, query: 'rails'
      expect(response).to have_http_status(:success)
      expect(response).to render_template(partial: 'autocomplete/_repository_suggestions')
    end
  end
end
```

I don't know if it's a proper way to test this controller.

Could you please give me some advice?

Thanks! :smile:

tarebyte commented:
@cyhsutw sorry for not being more clear, I meant test that a known search gives you back the results you expect.

Like `rails/rails` returns the ruby on rails project.

johndbritton commented on app/controllers/autocomplete_controller.rb:
```diff
@@ -0,0 +1,33 @@
+
+  private
+
+  def required_scopes
```
How does this behave if we don't have the right scopes? Since we're requesting this endpoint via ajax, there's no opportunity to ask for more scopes.

cyhsutw replied:
@johndbritton 

Currently, there isn't any error handling in this controller (which is bad), I will try to add some of them to make it more robust.

johndbritton replied:
the `#required_scopes` method is used to request more scopes for the user if necessary. In this case it doens't really make sense, since it's happening in the context of an ajax request.

cyhsutw replied:
Sorry for misunderstanding the meaning of the method. I'll pushed a fix later tonight. 

Thank you for pointing this out.

cyhsutw commented:
@johndbritton @tarebyte 

Thank you very much for taking time to review my code :smile: 

I've update my code based on your suggestions.

Short summary:

- Add some specs to test the functionality of `AutocompleteController`. 
- Ability to pass/handle `repo_id` when using autocomplete (as suggested [here](https://github.com/education/classroom/pull/508#discussion-diff-56806759R10))
- Remove misuse of `#required_scopes`.

cyhsutw commented:
> johndbritton [had a problem deploying](https://heaven.githubapp.com/apps/classroom/deployments/539749/log) to production 23 hours ago

@johndbritton I saw this message in this PR, I'd like to know if there's anything I can help with?
(I've check the [log](https://heaven.githubapp.com/apps/classroom/deployments/539749/log) but it shows `Not Found`)

tarebyte commented:
@cyhsutw that's because the log it GitHub employee only 😄 

We were trying to see if we could use our internal tools to branch deploy a fork, it doesn't look like we can though

cyhsutw commented:
@tarebyte Thanks for the explanation! :smile:

tarebyte commented:
@cyhsutw can you merge master back in one more time? I think this is good to 🚢

johndbritton commented on spec/controllers/assignments_controller_spec.rb:
```diff
@@ -66,6 +66,23 @@
       end
     end
+
+    context 'repo_id for starter_code is passed' do
```
You need one more context here `invalid repo_id for starter_code is passed`, this could be the case of a repository the user doesn't have access to or an invalid input for `repo_id` like the string `abc` or `0`.

johndbritton commented on spec/controllers/autocomplete_controller_spec.rb:
```diff
@@ -0,0 +1,38 @@
+  end
+
+  describe 'GET #github_repos', :vcr do
+    it 'returns success and returns a subset of repos of the user' do
```
best practice here is to have a single `expect` per `it` block. That way the test output is more descriptive when something fails.

johndbritton replied:
So you could split this up into multiple `it` blocks and extract any repeated setup to a `before` block.

johndbritton commented on spec/controllers/autocomplete_controller_spec.rb:
```diff
@@ -0,0 +1,38 @@
+      expect(response).to render_template(partial: 'autocomplete/_repository_suggestions')
+    end
+
+    it 'returns success and returns search results of a query' do
```
Same thing here, you can split up your `expect`s into separate `it` blocks.

johndbritton commented:
@cyhsutw Left a few more inline notes, but this is looking very good. Once those last few things are addressed I think this is ready to deploy.

cyhsutw commented:
@johndbritton @tarebyte Thanks for spending time reviewing my code.

I've updated the code based on your suggestions.

> You need one more context here invalid repo_id for starter_code is passed, this could be the case of a repository the user doesn't have access to or an invalid input for `repo_id` like the string `abc` or `0`.

Actually, the controller DOES create assignments with those invalid `repo_id`s!
To address this, I added a [method](https://github.com/education/classroom/pull/508/files#diff-063449d33fb1ea54624f576bba84869f) in `concerns/starter_code.rb` to check if the user has proper access to that repo.

johndbritton commented:
@cyhsutw Thank you for your contribution, this has been deployed to production.

cyhsutw commented:
@johndbritton @tarebyte Thank you for taking time reviewing my code, helped me a lot!
